### PR TITLE
Receive bug and more tests

### PR DESCRIPTION
**Pull request by LianaHus**

### CODE REVIEW

File: src/app/ui/universal-dapp-ui.js [hunk 1]
```diff
@@ -161,7 +161,7 @@ UniversalDAppUI.prototype.renderInstanceFromABI = function (contractABI, address
     const fallback = self.udapp.getFallbackInterface(contractABI)
     const receive = self.udapp.getReceiveInterface(contractABI)
     const args = {
-      funABI: fallback,
+      funABI: (fallback ? fallback : receive),
```
Comment (yann300) [inline src/app/ui/universal-dapp-ui.js]: and if both are defined?

Comment (LianaHus) [reply src/app/ui/universal-dapp-ui.js]: then it's fine. fallback is default one

File: src/app/ui/universal-dapp-ui.js [hunk 2]
```diff
@@ -177,16 +177,22 @@ UniversalDAppUI.prototype.renderInstanceFromABI = function (contractABI, address
       }
     }
     if (calldata) {
-      if (calldata.length > 3 && calldata.substr(0, 2) === '0x') {
+      if (calldata.length < 2) setLLIError('the calldata should be a valid hexadecimal value with size of at least one byte.')
+      if (calldata.length < 4 && calldata.substr(0, 2) === '0x') {
```
Comment (yann300) [inline src/app/ui/universal-dapp-ui.js]: this is more than that, the size should be even

Comment (LianaHus) [reply src/app/ui/universal-dapp-ui.js]: :) thanks

File: src/app/ui/universal-dapp-ui.js [hunk 3]
```diff
@@ -191,7 +191,17 @@ UniversalDAppUI.prototype.renderInstanceFromABI = function (contractABI, address
...
     }
-    if (!receive && !fallback) setLLIError("Both 'receive' and 'fallback' functions are not defined")
+
+    if (!receive && !fallback) setLLIError(`Both 'receive' and 'fallback' functions are not defined`)
+
+    // we have to put the right function ABI:
+    // if receive is defined and that there is no calldata => receive function is called
+    // if fallback is defined => fallback function is called
+    if (receive && !calldata) args.funABI = receive
+    if (fallback) args.funABI = fallback
```
Comment (LianaHus) [inline src/app/ui/universal-dapp-ui.js]: this will rewrite your previous change

Comment (yann300) [reply src/app/ui/universal-dapp-ui.js]: indeed , changed it

**Pull request merged**